File: src/test/java/com/mikea/bayes/sampling/ForwardSamplerTest.java
package com.mikea.bayes.sampling;

import com.mikea.bayes.BayesianNetwork;
import com.mikea.bayes.Var;
import com.mikea.bayes.VarAssignment;
import com.mikea.bayes.data.StudentsNetwork;
import com.mikea.bayes.query.QueryAlgorithm;
import org.junit.Test;

import static junit.framework.Assert.assertEquals;

/**
 * @author mike.aizatsky@gmail.com
 */
public class ForwardSamplerTest {
    @Test
    public void testProbabilities() throws Exception {
        BayesianNetwork network = StudentsNetwork.buildStudentsNetwork();
        int nSamples = 10000;

        QueryAlgorithm.Result queryResult = QueryAlgorithm.VAR_ELIMINATION.prepare(network);

        VarAssignment[] samples = ForwardSampler.sample(network, nSamples);

        int[][] counts = new int[network.V()][];
        for (VarAssignment sample : samples) {
            for (Var var : sample) {
                int v = var.getIndex();
                if (counts[v] == null) counts[v] = new int[var.getCardinality()];
                counts[v][sample.get(var)]++;
            }
        }

        for (int i = 0; i < network.V(); ++i) {
            Var var = network.getVar(i);
            int sum = 0;
            int[] varCounts = counts[var.getIndex()];
            for (int varCount : varCounts) {
                sum += varCount;
            }
            assertEquals(nSamples, sum);

            for (int j = 0; j < varCounts.length; j++) {
                double p = (double) varCounts[j] / sum;
                double expectedProb = queryResult.getProbability(new VarAssignment(new Var[]{var}, new int[]{j}));
                assertEquals(expectedProb, p, 1e-2);
            }

        }
    }
}

File: src/test/java/com/mikea/bayes/sampling/ForwardSamplerTest.scala
package com.mikea.bayes.sampling

import org.junit.Test

/**
 * @author mike.aizatsky@gmail.com
 */
class ForwardSamplerTest {
  @Test def testProbabilities {
    ???
/*
    val network: BayesianNetwork = StudentsNetwork.buildStudentsNetwork
    val nSamples: Int = 10000
    val queryResult: Nothing = QueryAlgorithm.VAR_ELIMINATION.prepare(network)
    val samples: Array[VarAssignment] = ForwardSampler.sample(network, nSamples)
    val counts: Array[Array[Int]] = new Array[Array[Int]](network.V)
    for (sample <- samples) {
      import scala.collection.JavaConversions._
      for (`var` <- sample) {
        val v: Int = `var`.getIndex
        if (counts(v) == null) counts(v) = new Array[Int](`var`.cardinality)
        counts(v)(sample.get(`var`)) += 1
      }
    }
    {
      var i: Int = 0
      while (i < network.V) {
        {
          val `var`: Var = network.getVar(i)
          var sum: Int = 0
          val varCounts: Array[Int] = counts(`var`.getIndex)
          for (varCount <- varCounts) {
            sum += varCount
          }
          assertEquals(nSamples, sum)
          {
            var j: Int = 0
            while (j < varCounts.length) {
              {
                val p: Double = varCounts(j).asInstanceOf[Double] / sum
                val expectedProb: Double = queryResult.getProbability(new VarAssignment(Array[Var](`var`), Array[Int](j)))
                assertEquals(expectedProb, p, 1e-2)
              }
              ({
                j += 1; j - 1
              })
            }
          }
        }
        ({
          i += 1; i
        })
      }
    }
*/
  }
}